File: ramdisk_tools/ioflash/IOFlashPartitionScheme.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include "ioflash.h"
#include "IOFlashPartitionScheme.h"

//http://en.wikipedia.org/wiki/Fletcher's_checksum
uint16_t fletcher16( uint8_t const *data, size_t bytes )
{
    uint16_t sum1 = 0xff, sum2 = 0xff;

    while (bytes) {
            size_t tlen = bytes > 20 ? 20 : bytes;
            bytes -= tlen;
            do {
                    sum2 += sum1 += *data++;
            } while (--tlen);
            sum1 = (sum1 & 0xff) + (sum1 >> 8);
            sum2 = (sum2 & 0xff) + (sum2 >> 8);
    }
    /* Second reduction step to reduce sums to 8 bits */
    sum1 = (sum1 & 0xff) + (sum1 >> 8);
    sum2 = (sum2 & 0xff) + (sum2 >> 8);
    return sum2 << 8 | sum1;
}

IOFlashPartitionScheme* IOFlashPartitionScheme_init(IOFlashController_client* iofc, uint8_t* buffer)
{
    struct IOFlashPartitionTable* p = (struct IOFlashPartitionTable*) buffer;
    if (p->magic != 'Grdn')
    {
        fprintf(stderr, "Bad Grdn magic %x\n", p->magic);
        return NULL;
    }
    //TODO: check checksum1

    if (p->cemetary.magic != CEMETARY_GATE)
    {
        fprintf(stderr, "Bad cemetary gate %x\n", p->cemetary.magic);
        return NULL;
    }
    if (fletcher16((uint8_t*) &p->cemetary, 42) != p->cemetary.checksum)
    {
        fprintf(stderr, "Bad cemetary checksum\n");
        return NULL;
    }

    if (p->spares.magic != 'spAr')
    {
        fprintf(stderr, "Bad rAps magic %x\n", p->spares.magic);
        return NULL;
    }
    if (fletcher16((uint8_t*) &p->spares, 0x10A) != p->spares.checksum)
    {
        fprintf(stderr, "Bad spares checksum\n");
        return NULL;
    }

    if (p->subs.magic != 'subs')
    {
        fprintf(stderr, "Bad sbus magic %x\n", p->subs.magic);
        return NULL;
    }
    IOFlashPartitionScheme* fps = (IOFlashPartitionScheme*) malloc(sizeof(IOFlashPartitionScheme));

    fps->flash = iofc;
    fps->ceCount = iofc->num_ce;

    memcpy((void*) &fps->ptable, (void*) p, sizeof(struct IOFlashPartitionTable));

    uint32_t i, blk_offset;

    for(i=0,blk_offset=0; i < 32; i++)
    {
        if( fps->ptable.partitions[i].flags & kIOFlashPartitionSchemePool)
        {
            fps->ptable.partitions[i].num_blocks = (fps->ptable.partitions[i].start_block * fps->ptable.partitions[i].num_blocks) / fps->ceCount;
            fps->ptable.partitions[i].start_block = blk_offset;
        }
        blk_offset += fps->ptable.partitions[i].num_blocks;
    }

    return fps;
}

uint32_t get_partition_idx_for_block(IOFlashPartitionScheme* fps, uint32_t block)
{
    uint32_t i;
    for(i=0; i < 32; i++)
    {
        if( block >= fps->ptable.partitions[i].start_block
        && block < fps->ptable.partitions[i].start_block + fps->ptable.partitions[i].num_blocks)
        {
            return i;
        }
    }
    return -1;
}


uint32_t IOFlashPartitionScheme_get_flags_for_block(IOFlashPartitionScheme* fps, uint32_t ce, uint32_t block)
{
    uint32_t i,idx;
    uint32_t original_block;

    //uint32_t vblock = block * fps->ceCount + ce;
    //check if bad boot block
    /*if (block < 16)
    {
        uint8_t x = fps->ptable.cemetary.cemetary[vblock / 8];
        return (x & (1 << (vblock % 8))) == 0;
    }*/
    uint32_t sp = (block & 0xFFFFFF) | (ce << 24);
    //check if spare
    //printf("sp=%x\n", sp);
    for (i=0; i < fps->ptable.spares.num_spares; i++)
    {
        if (fps->ptable.spares.spares[i] == sp)
        {
            //check if spare is "in use"
            original_block = fps->ptable.subs.subs[i];
            if (original_block != 0)
            {
                idx = get_partition_idx_for_block(fps, original_block/fps->ceCount);
                if(idx != -1)
                {
                    //printf("lol %x %d %x\n", original_block, idx, ptable.partitions[idx].flags);
                    //return (fps->ptable.partitions[idx].flags & kIOFlashPartitionSchemeUseFullPages) == 0;
                    return fps->ptable.partitions[idx].flags;
                }
            }
        }
    }
    if (block >= 16)//HAX
        return kIOFlashPartitionSchemeUseFullPages;
    idx = get_partition_idx_for_block(fps, block);
    if(idx != -1)
        return fps->ptable.partitions[idx].flags;
    return 0;
}

uint32_t IOFlashPartitionScheme_remap_bootloader_block(IOFlashPartitionScheme* fps, uint32_t ce, uint32_t block, uint32_t flags, uint32_t* realCE, uint32_t* realBlock)
{
    uint32_t i;
    uint32_t vblock = block * fps->ceCount + ce;
    uint8_t x = fps->ptable.cemetary.cemetary[vblock / 8];

    uint32_t bad = (x & (1 << (vblock % 8)));

    if (!bad && !(flags & kIOFlashPartitionSchemePool))
    {
        *realCE = ce;
        *realBlock = block;
        return 0;
    }
    for (i=0; i < fps->ptable.subs.num_subs; i++)
    {
        if (fps->ptable.subs.subs[i] == vblock)
        {
            uint32_t spare = fps->ptable.spares.spares[i];
            *realCE = spare >> 24;
            *realBlock = spare & 0xFFFFFF;
            return 0;
        }
    }
    fprintf(stderr, "remap_bootloader_block: cannot find sub for ce %d block %d\n", ce, block);
    return -1;
}

uint32_t get_partition_idx(IOFlashPartitionScheme* fps, const char* name)
{
    uint32_t i;
    for(i=0; i < 32; i++)
    {
        if(fps->ptable.partitions[i].name == *((uint32_t*)name))
            return i;
    }
    return -1;
}


uint32_t IOFlashPartitionScheme_read_partition(IOFlashPartitionScheme* fps, const char* name, uint8_t** buffer, uint32_t* size)
{
    IOFlashControllerUserClient_OutputStruct out;
    uint32_t ce, block, page, off, pageAddr, remappedCE, remappedBlock;
    uint32_t pageSize, metaPerLogicalPage, partitionSize, pagesPerBlock;
    uint32_t page_bits, block_bits, cau_bits;
    uint32_t is_ppn, slc;
    IOReturn r;

    uint32_t partition_idx = get_partition_idx(fps, name);
    if (partition_idx == -1)
    {
        fprintf(stderr, "read_partition: cant find partition %s\n", name);
        return -1;
    }
    struct IOFlashPartitionDesc* partition = &fps->ptable.partitions[partition_idx];

    if (partition->flags & kIOFlashPartitionSchemeUseFullPages)
    {
        pageSize = fps->flash->page_bytes;
    }
    else
    {
        pageSize = fps->flash->bootloader_bytes;
    }
    metaPerLogicalPage = fps->flash->meta_per_logical_page;

    //XXX we only read in first CE
    ce = 0;

    slc = (partition->flags & kIOFlashPartitionSchemeUseSLCBlocks) != 0;
    page_bits = fps->flash->page_bits;
    block_bits = fps->flash->block_bits;
    cau_bits = fps->flash->cau_bits;
    is_ppn = fps->flash->ppn_device;

    pagesPerBlock = fps->flash->block_pages;

    if (partition->flags & kIOFlashPartitionSchemeUseSLCBlocks)
    {
        pagesPerBlock = fps->flash->slc_pages; //TODO get param
    }

    partitionSize = partition->num_blocks * pagesPerBlock * (pageSize);

    uint8_t* data = valloc(partitionSize);
    if( data == NULL)
    {
        fprintf(stderr, "read_partition: cant valloc %d bytes\n", partitionSize);
        return -1;
    }

    uint8_t* pageBuffer = valloc(pageSize);
    if( pageBuffer == NULL)
    {
        fprintf(stderr, "read_partition: cant valloc %d bytes\n", pageSize);
        return -1;
    }
    uint8_t* spareBuffer = valloc(metaPerLogicalPage);
    if( spareBuffer == NULL)
    {
        fprintf(stderr, "read_partition: cant valloc %d bytes\n", metaPerLogicalPage);
        return -1;
    }

    fprintf(stderr, "read_partition is_ppn %d pagesPerBlock %d pageSize %d slc %d\n", is_ppn, pagesPerBlock, pageSize, slc);

    for(off=0, block=partition->start_block; block < (partition->start_block + partition->num_blocks); block++)
    {
        IOFlashPartitionScheme_remap_bootloader_block(fps, ce, block, partition->flags, &remappedCE, &remappedBlock);

        for(page=0; page < pagesPerBlock; page++)
        {
            if (is_ppn)
            {
                pageAddr = (remappedBlock << page_bits) | page | (slc << (page_bits+block_bits+cau_bits));
            }
            else
            {
                pageAddr = remappedBlock*pagesPerBlock + page;
            }

            if (partition->flags & kIOFlashPartitionSchemeUseFullPages)
            {
                r = FSDReadPageWithOptions(fps->flash,
                                       remappedCE,
                                       pageAddr,
                                       pageBuffer,
                                       spareBuffer,
                                       metaPerLogicalPage,
                                       1,
                                       &out);
                if(r == 0 && spareBuffer[0] != (0x40 | (partition_idx & 0xF)))
                {
                    //fprintf(stderr, "skipping weird page at %d %d %x\n", remappedBlock, page, spareBuffer[0]);
                    continue;
                }
            }
            else
            {
                r = FSDReadBootPage(fps->flash, remappedCE, pageAddr, pageBuffer, &out);
            }

            memcpy(&data[off], pageBuffer, pageSize);
            off += pageSize;
        }
    }
    free(pageBuffer);
    free(spareBuffer);

    *buffer = data;
    *size = off;
    return 0;
}

File: python_scripts/nand/IOFlashPartitionScheme.py
import struct

#from http://en.wikipedia.org/wiki/Fletcher's_checksum
def fletcher16(data):
    sum1 = 0xff
    sum2 = 0xff
    _bytes = len(data)
    i = 0

    while (_bytes):
        if _bytes > 20: tlen = 20
        else: tlen = _bytes
        _bytes -= tlen
        for _ in xrange(tlen):
            sum1 += ord(data[i])
            sum2 += sum1
            i += 1
        sum1 = (sum1 & 0xff) + (sum1 >> 8)
        sum2 = (sum2 & 0xff) + (sum2 >> 8)
    #/* Second reduction step to reduce sums to 8 bits */
    sum1 = (sum1 & 0xff) + (sum1 >> 8)
    sum2 = (sum2 & 0xff) + (sum2 >> 8)
    return sum2 << 8 | sum1;

"""
ndrG 
0x4 should_be_zero ;)
0x8 versionMajor
0xC versionMinor = 6 (ios4/5) 9 (ios6)
0x10 generation
0x14 checksumXXX
0x20 factoryBBT[0x1E0] : last_block num_bad + [59* 8 bytes entries]
0x200 ???
0x220
0x224 0xDEADCAFE => cemetary
0x250 spares
0x35c subs
0x400 partition table
0x600 end

boot
2*0x600 ptab
LLB

---
partition table Diff
stored in slack space in nvrm (todo) & plog (closing pages)
ffiD
uint16_t size
uint16_t sequence
uint32_t major  (must match ptable)
uint32_t minor  (must match ptable)
uint32_t generation (must match ptable)
0x20 => up-to-date cemetary

spare
0x40 | (part_idx & 0xF) | 0x0 
"""

kIOFlashPartitionSchemeFakePart = 1 #set for scfg, fbbt, diag that arent managed ?
kIOFlashPartitionSchemeIsSclice     = 2 #wut, new in ios6?
kIOFlashPartitionSchemeIsPool     = 8 #wut, new in ios6?
kIOFlashPartitionSchemeUseSLCBlocks  = 0x100 #sometimes set for plog/nvrm, only if ppn-device
kIOFlashPartitionSchemeUseFullPages  = 0x200 #sometimes set for firm


#TODO: handle partition table Diff
class IOFlashPartitionScheme(object):
    def __init__(self, nand, data):
        self.nand = nand
        self.pagesPerBlock = nand.pagesPerBlock
        self.nCEs = nand.nCEs
        self.cemetary = []
        self.spares = []
        self.subs = []
        self.parts = {}
        self.parts_names = []

        self.validatePartitionTable(data)
        if not self.initCemetery(data[0x224:0x224+42+2]):
            raise Exception("Cemetery is haunted :)")
        self.initSpares(data[0x250:0x35C])
        self.initSubs(data[0x35C:0x400])

        blk_offset = 0
        for idx in xrange(0x400,0x600,0x10):
            name = data[idx:idx+4][::-1]
            if name == "none":
                continue
            a,b,flags = struct.unpack("<LLL", data[idx+4:idx+16])
            if flags & kIOFlashPartitionSchemeIsPool:
                b = (a*b) / self.nCEs
                a = blk_offset
            self.parts[name] = a,b,flags
            blk_offset += b
            self.parts_names.append(name)

    def show(self):
        print "".join(map(lambda x:x.ljust(12), ["Name", "Start block", "Size", "Flags"]))
        for name in self.parts_names:
            start, size, flags = self.parts[name]
            print "".join(map(lambda x:str(x).ljust(12), [name, start, size, "0x%x" % flags]))

    def probe(self, data):
        return False

    def validatePartitionTable(self, data):
        if data[:4] != "ndrG":
            return False
        major = struct.unpack(">L", data[8:12])[0]
        return major == 0

    def initCemetery(self, data):
        gate = struct.unpack("<L", data[0:4])[0]
        if gate != 0xdeadcafe:
            print "Bad cemetery gate %x" % gate
            return False
        guards = struct.unpack("<H", data[42:44])[0]
        if guards != fletcher16(data[:42]):
            print "Bad cemetery guards %x" % guards
            return False
        self.cemetary = map(ord, data[10:42])
        return True

    def initSpares(self, data):
        if data[:4] != "rAps":
            print "Bad spares magic"
            return
        x = fletcher16(data[:0x10A])
        guards = struct.unpack("<H", data[0x10A:0x10C])[0]
        if x != guards:
            print "Bad spares checksum"
            return
        n_spares = struct.unpack("<H", data[4:6])[0]
        assert n_spares < 64
        idx =  10
        for idx in xrange(10, 10+ n_spares*4, 4):
            block = struct.unpack("<L", data[idx:idx+3] + "\x00")[0]
            ce = ord(data[idx+3:idx+4])
            self.spares.append((ce,block))

    def initSubs(self, data):
        if data[:4] != "sbus":
            print "Bad subs magic"
            return
        n_subs = struct.unpack("<H", data[4:6])[0]
        assert n_subs < 64
        self.subs = map(ord, data[10:10+n_subs])

    def isMappedPartition(self, idx):
        n = self.parts_names[idx]
        return n in ["plog","nvrm","firm"]

    def is_bad_vblock(self, vblock):
        x = self.cemetary[vblock / 8]
        return (x & (1 << (vblock % 8))) != 0

    def remap_block(self, ce, block):
        vblock = block * self.nCEs + ce
        try:
            sub = self.subs.index(vblock)
            ce, block = self.spares[sub]
        except:
            pass
        return ce, block, vblock

    def readPartitionBlock(self, name, ce, block):
        if not name in ["boot", "plog","nvrm","firm"]:
            return
        start, end, flags = self.parts[name]
        partition_idx = self.parts_names.index(name)
        block += start

        vblock = block * self.nCEs + ce
        if self.is_bad_vblock(vblock):
            assert vblock in self.subs
            sub = self.subs.index(vblock)
            ce, block = self.spares[sub]
            print "Reading remapped bootloader block %s %d => %d:%d !" % (name, vblock, ce,block)
        elif flags & kIOFlashPartitionSchemeIsPool:
            #print "Partition %s has pool flag, remapping" % name
            sub = self.subs.index(vblock)
            #print "sub ce=%d block=%d vblock=%d sub=%d" % (ce, block, vblock, sub)
            ce, block = self.spares[sub]
            #print "=> ce=%d block=%d" % (ce,block)

        pagesPerBlock = self.pagesPerBlock
        if flags & kIOFlashPartitionSchemeUseSLCBlocks:
            pagesPerBlock = self.nand.slc_pages
            page_bits = self.nand.page_bits
            bits = self.nand.block_bits + self.nand.cau_bits + page_bits
        #print name, block,self.pagesPerBlock, "%x"  %flags
        res = ""
        for i in xrange(pagesPerBlock):
            if flags & kIOFlashPartitionSchemeUseSLCBlocks:
                pageaddr = i | block << page_bits | 1 << bits
            else:
                pageaddr = block*self.pagesPerBlock + i

            if flags & kIOFlashPartitionSchemeUseFullPages: #!!!
                spare, data = self.nand.readPage(ce, pageaddr, spareType=None)
                if spare and ord(spare[0]) ==  0x40 | (partition_idx & 0xF):
                    res += data
                elif spare:
                    print "skipping weird page at %d %d %x" % (block, i, ord(spare[0]))
            else:
                res += self.nand.readBootPage(ce, pageaddr)
        return res

    def readPartition(self, name):
        if not name in ["boot", "plog","nvrm","firm"]:
            return
        start, end, flags = self.parts[name]
        res = ""
        for b in xrange(end):
            res += self.readPartitionBlock(name, 0, b)
        return res

